### PR TITLE
[ONNX] Symbolic: dynamic input for OneHot, bool for Einsum

### PR DESCRIPTION
**Pull request by jiafatom**
Fixes zcode-moe training export.

### CODE REVIEW

File: torch/onnx/symbolic_opset9.py [hunk 1]
```diff
@@ -2666,11 +2666,12 @@ def prim_tolist(g, input, dim_val, elem_ty_val):
...
 
 
-@parse_args('v', 'i')
 def one_hot(g, self, num_classes):
     values = g.op("Constant", value_t=torch.LongTensor([0, 1]))
-    depth = g.op("Constant", value_t=torch.LongTensor([num_classes]))
-    return g.op("OneHot", self, depth, values, axis_i=-1)
+    # onnxruntime supports limited type combinations for OneHot.
+    if num_classes.type().scalarType() in ("Byte", "Char", "Int", "Short"):
+        num_classes = g.op("Cast", num_classes, to_i=sym_help.cast_pytorch_to_onnx["Long"])
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py:2673]: curious, will "Cast" be folded if `num_classes` is constant?

Comment (jiafatom) [reply torch/onnx/symbolic_opset9.py:2673]: It should fold, const folding https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/onnx/constant_fold.cpp#L402-L409 takes care of it.

File: torch/onnx/symbolic_opset12.py [hunk 2]
```diff
@@ -12,17 +12,26 @@
 
 # This file exports ONNX ops for opset 12
 
+def einsum_helper(g, equation, tensors):
+    if not tensors:
+        raise RuntimeError("Einsum inputs are empty.")
+    if tensors[0].type().scalarType() == "Bool":
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset12.py:18]: is this because onnxruntime does not support Bool for einsum? If so please comment it out too.

**Review by BowenBao — APPROVED** [review]
LGTM, minor suggestion with comment

File: torch/onnx/symbolic_opset12.py (hunk 2 above)
Comment (jiafatom) [reply torch/onnx/symbolic_opset12.py:19]: done

**Pull request merged**